File: app/src/test/java/com/example/mike/mp3player/service/library/content/searcher/ContentResolverSearcherTestBase.java
package com.example.mike.mp3player.service.library.content.searcher;

import android.content.ContentResolver;
import android.database.Cursor;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.MediaBrowserCompat.MediaItem;

import com.example.mike.mp3player.service.library.content.parser.ResultsParser;
import com.example.mike.mp3player.service.library.search.SearchDao;
import com.example.mike.mp3player.service.library.search.SearchEntity;

import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public abstract class ContentResolverSearcherTestBase<T extends ContentResolverSearcher> {

    T searcher;

    static final String idPrefix = "idPrefix";

    static final String VALID_QUERY = "VALID_QUERY";
    static final String INVALID_QUERY = "INVALID_QUERY";

    @Mock
    ContentResolver contentResolver;

    @Mock
    Cursor cursor;

    @Mock
    ResultsParser resultsParser;

    static List<MediaItem> expectedResult = new ArrayList<>();


    static {
        expectedResult.add(mock(MediaItem.class));
    }

    public abstract void testGetMediaType();

    public abstract void testSearchValidMultipleArguments();

    @Test
    public void testSearchInvalid() {
        when(searcher.resultsParser.create(eq(any()), idPrefix)).thenReturn(expectedResult);
        List<?> result = searcher.search(INVALID_QUERY);
        assertNotEquals(expectedResult, result);
    }

}

File: service/src/testFullDebug/java/com/github/goldy1992/mp3player/service/library/content/searcher/ContentResolverSearcherTestBase.kt
package com.github.goldy1992.mp3player.service.library.content.searcher

import android.content.ContentResolver
import android.database.Cursor
import androidx.media3.common.MediaItem
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.StandardTestDispatcher
import kotlinx.coroutines.test.TestCoroutineScheduler
import kotlinx.coroutines.test.TestScope
import kotlinx.coroutines.test.runTest
import org.junit.Assert
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

@OptIn(ExperimentalCoroutinesApi::class)
abstract class ContentResolverSearcherTestBase<T : ContentResolverSearcher<*>?> {
    var searcher: T? = null
    var idPrefix: String? = null

    var contentResolver: ContentResolver = mock<ContentResolver>()

    var cursor: Cursor = mock<Cursor>()

    private val testScheduler = TestCoroutineScheduler()
    protected val dispatcher  = StandardTestDispatcher(testScheduler)
    protected val testScope = TestScope(dispatcher)

    companion object {
        const val VALID_QUERY = "VALID_QUERY"
        const val INVALID_QUERY = "INVALID_QUERY"
        var expectedResult: MutableList<MediaItem> = ArrayList()

        init {
            expectedResult.add(mock<MediaItem>())
        }
    }

    abstract fun testGetMediaType()
    abstract fun testSearchValidMultipleArguments()
    @Test
    fun testSearchInvalid() = runTest(dispatcher)  {
        whenever(searcher!!.resultsParser.create(eq(any<Cursor>()), idPrefix!!)).thenReturn(expectedResult)
        var result: List<*>? = searcher!!.search(INVALID_QUERY)
        Assert.assertNotEquals(expectedResult, result)
    }
}